flask/app/groups.py

```diff
@@ -0,0 +1,230 @@
+from flask import abort, jsonify, request, Response, current_app as app
+from flask_login import login_user, logout_user, current_user, login_required
+from app import db, login, models
+from sqlalchemy.orm import contains_eager, joinedload
+from .routes import check_admin, transaction_or_abort, mixin
+from minio import Minio
+from .madmin import MinioAdmin, readwrite_buckets_policy
+
+
+@app.route("/api/groups", methods=["GET"])
+@login_required
+def list_groups():
+    if app.config.get("LOGIN_DISABLED") or current_user.is_admin:
+        user_id = request.args.get("user")
+    else:
+        user_id = current_user.user_id
+
+    if user_id:
+        groups = (
+            models.Group.query.join(models.Group.users)
+            .filter(models.User.user_id == user_id)
+            .all()
+        )
+    else:
+        groups = models.Group.query.all()
+
+    return jsonify(
+        [
+            {"group_code": group.group_code, "group_name": group.group_name}
+            for group in groups
+        ]
+    )
+
+
+@app.route("/api/groups/<string:group_code>", methods=["GET"])
+@login_required
+def get_group(group_code):
+    if app.config.get("LOGIN_DISABLED") or current_user.is_admin:
+        user_id = request.args.get("user")
+    else:
+        user_id = current_user.user_id
+
+    if user_id:
+        group = (
+            models.Group.query.filter(models.Group.group_code == group_code)
+            .join(models.Group.users)
+            .filter(models.User.user_id == user_id)
+            .one_or_none()
+        )
+    else:
+        group = models.Group.query.filter(
+            models.Group.group_code == group_code
+        ).one_or_none()
+
+    if group is None:
+        return "Forbidden", 403
+
+    return jsonify(
+        {
+            "group_code": group.group_code,
+            "group_name": group.group_name,
+            "users": [
+                user.username for user in group.users if user.deactivated == False
+            ],
+        }
+    )
+
+
+@app.route("/api/groups/<string:group_code>", methods=["PATCH"])
+@login_required
+@check_admin
+def update_group(group_code):
+
+    if not request.json:
+        return "Request body must be JSON", 415
+
+    group = models.Group.query.filter_by(group_code=group_code).first_or_404()
+
+    if "group_name" in request.json:
+        if request.json["group_name"]:
+            # Check if display name is in use, 422
+            conflicting_group = models.Group.query.filter_by(
+                group_name=request.json["group_name"]
+            ).one_or_none()
+            if (conflicting_group is not None) and (conflicting_group != group):
+                return "Group name in use", 422
+            group.group_name = request.json["group_name"]
+
+    minioAdmin = MinioAdmin(
```

Comment:
Please use snake_case since this is in Python

Reply:
Old habits lol